File: ProducerConsumer/app/src/main/java/ProdConsStream/DataCleaner.java
package ProdConsStream;

import unibo.actors23.ActorBasic23;
import unibo.actors23.ActorContext23;
import unibo.basicomm23.interfaces.IApplMessage;
import unibo.basicomm23.utils.CommUtils;

public class DataCleaner extends ActorBasic23 {
    public DataCleaner(String name, ActorContext23 ctx) {
        super(name, ctx);
        subscribeLocalActor("producer");
    }

    @Override
    protected void elabMsg(IApplMessage msg) throws Exception {
        //CommUtils.outgray(name + "msg=" + msg);
        int n = Integer.parseInt( msg.msgContent() );
        if( n % 2 == 0 ) { //elimino i dispari
            IApplMessage  event = CommUtils.buildEvent(name, "data", "" + n);
            emitLocalStreamEvent(event);
        }
    }
}

File: qakdemo23/resources/rx/dataCleaner.kt
package rx
 

import it.unibo.kactor.*
import kotlinx.coroutines.delay
import it.unibo.kactor.ActorBasic
import alice.tuprolog.Term
import alice.tuprolog.Struct
import unibo.basicomm23.interfaces.IApplMessage
 

 
class dataCleaner (name : String ) : ActorBasic( name ) {
val LimitLow  = 2	
val LimitHigh = 150
 
    override suspend fun actorBody(msg: IApplMessage) {
  		elabData( msg )
 	}

 	
 
	  suspend fun elabData( msg: IApplMessage ){ //		 
 		val data  = (Term.createTerm( msg.msgContent() ) as Struct).getArg(0).toString()
  		//println("   $name |  data = $data ")		
		val Distance = Integer.parseInt( data ) 
 		if( Distance > LimitLow && Distance < LimitHigh ){
			emitLocalStreamEvent( msg ) //propagate
     	}else{
			//println("   $name |  DISCARDS $Distance ")
 		}				
 	}
}